File: CppPractice/Project1/Project1/Leetcode279.cpp
#include <string>

#include <vector>

#include <iostream>

#include <stdlib.h>

#include <time.h>

#include <sstream>

#include <algorithm>

#include <map>

#include <bitset>

using namespace std;


class Solution {
public:
    int numSquares(int n) {
        vector<int> dp(n + 1, INT_MAX);
        dp[0] = 0;
        for (int i = 1; i <= sqrt(n); i++)
        {
            dp[i * i] = 1;

        }
        for (int i = 0; i < n+1; i++)
        {
            for (int j = 1; j <= sqrt(i); j++)
            {
                dp[i] = min(dp[i], dp[i - j * j]+1);

            }

        }

        

        return dp.back();

    }
};

File: Pythonleetcode/Leetcode279.py
class Solution(object):
    def numSquares(self, n):
        ans=[n]*(n+1)
        ans[0] = 0
        ans[1]=1
        for i in range(2,n+1):
            j=1
            while(j*j<=i):
                ans[i] = min(ans[i], ans[i - j * j] + 1)

                j = j + 1

        return ans[n]





print(Solution().numSquares(13))